components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java

```diff
@@ -189,26 +183,87 @@ public String buildIDToken(OAuthTokenReqMessageContext tokenReqMsgCtxt,
         return OAuth2Util.signJWT(jwtClaimsSet, signatureAlgorithm, signingTenantDomain).serialize();
     }
 
-    private boolean isInvalidToken(JWTClaimsSet jwtClaimsSet) {
-        return !isValidIdToken(jwtClaimsSet);
-    }
+    @Override
+    public String buildIDToken(OAuthAuthzReqMessageContext authzReqMessageContext,
+                               OAuth2AuthorizeRespDTO tokenRespDTO) throws IdentityOAuth2Exception {
 
-    private long getIdTokenValidityInMillis() {
-        return OAuthServerConfiguration.getInstance().getOpenIDConnectIDTokenExpiryTimeInMillis();
-    }
+        String accessToken = tokenRespDTO.getAccessToken();
+        String clientId = authzReqMessageContext.getAuthorizationReqDTO().getConsumerKey();
+        String spTenantDomain = getSpTenantDomain(authzReqMessageContext);
+        String issuer = getIdTokenIssuer(spTenantDomain);
 
-    private boolean isEssentialClaim(AuthorizationGrantCacheEntry authorizationGrantCacheEntry,
-                                     String oidcClaimUri) {
-        return OAuth2Util.getEssentialClaims(authorizationGrantCacheEntry.getEssentialClaims(), OAuthConstants.ID_TOKEN)
-                .contains(oidcClaimUri);
+        // Get subject from Authenticated Subject Identifier
+        AuthenticatedUser authorizedUser = authzReqMessageContext.getAuthorizationReqDTO().getUser();
+        String subject = getSubjectClaim(authzReqMessageContext, clientId, spTenantDomain, authorizedUser);
+
+        String nonceValue = authzReqMessageContext.getAuthorizationReqDTO().getNonce();
+        LinkedHashSet acrValue = authzReqMessageContext.getAuthorizationReqDTO().getACRValues();
+
+        long idTokenLifeTimeInMillis = OAuthServerConfiguration.getInstance().getOpenIDConnectIDTokenExpiryTimeInMillis();
+        long currentTimeInMillis = Calendar.getInstance().getTimeInMillis();
+
+        if (log.isDebugEnabled()) {
+            log.debug(buildDebugMessage(issuer, subject, nonceValue, idTokenLifeTimeInMillis, currentTimeInMillis));
+        }
+
+        JWTClaimsSet jwtClaimsSet = new JWTClaimsSet();
+        jwtClaimsSet.setIssuer(issuer);
+
+        // Set the audience
+        List<String> audience = getOIDCAudience(clientId);
+        jwtClaimsSet.setAudience(audience);
+
+        jwtClaimsSet.setClaim(AZP, clientId);
+        jwtClaimsSet.setExpirationTime(getIdTokenExpiryInMillis(idTokenLifeTimeInMillis, currentTimeInMillis));
+        jwtClaimsSet.setIssueTime(new Date(currentTimeInMillis));
```

Comment:
Don't we need to use UTC time here? Calendar.getInstance(TimeZone.getTimeZone(UTC)

Reply:
new Data() does that internally i guess.